hazelcast/src/main/java/com/hazelcast/jet/impl/execution/ProcessorTasklet.java

```diff
@@ -720,12 +720,21 @@ public void provideDynamicMetrics(MetricDescriptor descriptor, MetricsCollection
             mContext.collect(descriptorWithOrdinal, EMITTED_COUNT, ProbeLevel.INFO, ProbeUnit.COUNT, emittedCounts.get(i));
         }
 
-        for (Byte key : coalescers.keys()) {
-            MetricDescriptor keyedDesc = descriptor.copy().withDiscriminator("key", Byte.toString(key));
-            mContext.collect(keyedDesc, TOP_OBSERVED_WM, ProbeLevel.INFO, ProbeUnit.MS, coalescers.topObservedWm(key));
-            mContext.collect(keyedDesc, COALESCED_WM, ProbeLevel.INFO, ProbeUnit.MS, coalescers.coalescedWm(key));
-            mContext.collect(keyedDesc, LAST_FORWARDED_WM, ProbeLevel.INFO, ProbeUnit.MS, outbox.lastForwardedWm(key));
-            mContext.collect(keyedDesc, LAST_FORWARDED_WM_LATENCY, ProbeLevel.INFO, ProbeUnit.MS, lastForwardedWmLatency(key));
+        if (!coalescers.keys().isEmpty()) {
+            for (Byte key : coalescers.keys()) {
+                MetricDescriptor keyedDesc = descriptor.copy().withDiscriminator("key", Byte.toString(key));
+                mContext.collect(keyedDesc, TOP_OBSERVED_WM, ProbeLevel.INFO, ProbeUnit.MS, coalescers.topObservedWm(key));
+                mContext.collect(keyedDesc, COALESCED_WM, ProbeLevel.INFO, ProbeUnit.MS, coalescers.coalescedWm(key));
+                mContext.collect(keyedDesc, LAST_FORWARDED_WM, ProbeLevel.INFO, ProbeUnit.MS, outbox.lastForwardedWm(key));
+                long lastForwardedWmLatency = lastForwardedWmLatency(key);
+                mContext.collect(keyedDesc, LAST_FORWARDED_WM_LATENCY, ProbeLevel.INFO, ProbeUnit.MS, lastForwardedWmLatency);
+            }
+        } else {
+            MetricDescriptor keyedDesc = descriptor.copy().withDiscriminator("key", "0");
+            mContext.collect(keyedDesc, TOP_OBSERVED_WM, ProbeLevel.INFO, ProbeUnit.MS, Long.MIN_VALUE);
+            mContext.collect(keyedDesc, COALESCED_WM, ProbeLevel.INFO, ProbeUnit.MS, Long.MIN_VALUE);
+            mContext.collect(keyedDesc, LAST_FORWARDED_WM, ProbeLevel.INFO, ProbeUnit.MS, Long.MIN_VALUE);
+            mContext.collect(keyedDesc, LAST_FORWARDED_WM_LATENCY, ProbeLevel.INFO, ProbeUnit.MS, 0L);
```

Comment:
Why do we need this values? Can't we just not collect anything if keys are empty?

Reply:
To rephrase I understand what this PR does and it's correct, but I don't understand why.

Reply:
@barisbasaran @zhmaeff knows more, it's MC side specifics.

Reply:
what do you think @zhmaeff ?

Reply:
Without these changes MC couldn't find some metrics that were present before. So it is just for preserving previous behavior from the MC perspective.

Reply:
Thanks.